### PR TITLE
Haptics stable updated

### PR DESCRIPTION
**Pull request by rayanisran**
This PR includes some changes to the Haply's force calculation to attempt to create a smoother transition from segment to segment, particularly for those to be demoed at CSUN. Also includes a fix for resetting device torques in case the device does go unstable.

### CODE REVIEW

File: src/info/info-utils.ts [hunk 1]
```diff
@@ -49,11 +53,11 @@ export function createButton(contentDiv: HTMLElement, id: string, text: string )
  * @param contentDiv container for canvas.
  * @returns Canvas with context.
  */
-export function createCanvas(contentDiv: HTMLElement, width:number, height:number){
+export function createCanvas(contentDiv: HTMLElement) {
```
Comment (jaydeepsingh25) [inline src/info/info-utils.ts:56]: I think it is changed by mistake, please keep this function unchanged

File: src/info/info.ts [hunk 2]
```diff
@@ -52,6 +53,8 @@ port.onMessage.addListener(async (message) => {
         renderings = { "request_uuid": request_uuid, "timestamp": 0, "renderings": [] };
     }
 
+    console.log(renderings);
```
Comment (jaydeepsingh25) [inline src/info/info.ts:56]: should be removed

File: src/info/info.ts [hunk 3]
```diff
@@ -243,12 +245,12 @@ port.onMessage.addListener(async (message) => {
             let btnNext = utils.createButton(contentDiv, "btnNext", "Next");
             let btnPrev = utils.createButton(contentDiv, "btnPrev", "Previous");
 
-            // creating canvas
-            const canvas= utils.createCanvas(contentDiv,canvasWidth,canvasHeight);
-
+            // creating canvas   
+            const canvas= utils.createCanvas(contentDiv);
```
Comment (jaydeepsingh25) [inline src/info/info.ts:249]: again, i think it is changed by mistake

File: src/info/info.ts [hunk 4]
```diff
@@ -260,6 +262,7 @@ port.onMessage.addListener(async (message) => {
 
             // world resolution properties
             const worldPixelWidth = 800;
+            const pixelsPerMeter = 6000;
```
Comment (jaydeepsingh25) [inline src/info/info.ts:265]: should be removed, as it is not used anywhere

File: src/info/info.ts [hunk 5]
```diff
@@ -212,14 +215,13 @@ port.onMessage.addListener(async (message) => {
             let deviceOrigin: Vector;
 
             // virtual end effector avatar offset
-            let firstCall:boolean = true;
+            const offset = 150;
+            let firstCall = true;
```
Comment (jaydeepsingh25) [inline src/info/info.ts:219]: keep the original changes

**Comment by sriGanna** [comment]
@jaydeepsingh25 we've reverted the old worker.ts and made manual changes so the change log should be more clear now.

File: src/info/worker.ts [hunk 6]
```diff
@@ -694,14 +644,89 @@ function activeGuidance(segments: SubSegment[][], tSegmentDuration: number,
...
+                              // force reduction as we approach the threshold
+                              let k = 3;
+                              let coeff = 1;
+                              if (diff.mag() < 0.025) {
+                                  let unit = diff.unit();
+                                  const multiplier = 1 / (diff.mag() + 0.9);
+                                  const coeff = Math.min(1, 1 - ((unit.x * unit.y) * multiplier * multiplier));
+                              }
+                              // move to new point with the WaitTime refresh rate
+                              if (Date.now() - tHoldTimeSegToSeg > tWaitTime) {
```
Comment (jaydeepsingh25) [inline src/info/worker.ts:683]: tWaitTime is undefined?

File: src/info/worker.ts [hunk 7]
```diff
@@ -683,6 +616,23 @@ function activeGuidance(segments: SubSegment[][], tSegmentDuration: number,
...
     else {
+      
+            // get the distance from where we currently are
+            const coord = currentSubSegment.coordinates[currentSubSegmentPointIndex];
+            const setPoint = new Vector(coord.x, coord.y);
+            const diff = setPoint.subtract(convPosEE.clone());
+
+            // don't set the time-based guidance until we have reached the first position
+            // set dist threshold
+            if (currentSubSegmentPointIndex == 0 && diff.mag() >= threshold) {
```
Comment (jaydeepsingh25) [inline src/info/worker.ts:627]: threshold is undefined?

File: src/info/worker.ts [hunk 8]
```diff
@@ -694,14 +644,89 @@ function activeGuidance(segments: SubSegment[][], tSegmentDuration: number,
...
 }
 
+enum Transition {
+  GetPoints,
+  Move,
+  Rest
+}
+
+// for transition between segments
+let transition: Transition = Transition.GetPoints;
```
Comment (jaydeepsingh25) [inline src/info/worker.ts:714]: i guess you need to move these variable declarations at the top somehere, I can see that they are being used in the lines above, where these will be undefined and the code may not work in an expected manner. Please check

File: src/info/info.ts [hunk 9]
```diff
@@ -389,6 +394,12 @@ port.onMessage.addListener(async (message) => {
             // event listener for serial comm button
             btn.addEventListener("click", async _ => {
                 // const worker = new Worker(browser.runtime.getURL("./info/worker.js"), { type: "module" });
+                
+                // only show the Arduino Zero
+                const filters = [
```
Comment (jaydeepsingh25) [inline src/info/info.ts:399]: Do we need this? I don't think it is used anywhere

File: src/info/worker.ts [hunk 10]
```diff
@@ -406,7 +339,7 @@ self.addEventListener("message", async function (event) {
           break;
       }
     }
-
+    const xHomeDiff = convPosEE.clone().subtract(xHome);
```
Comment (jaydeepsingh25) [inline src/info/worker.ts:342]: not used anywhere

File: src/info/worker.ts (hunk 6 above)
Comment (sriGanna) [reply src/info/worker.ts:698]: this should be defined at the top now as a const

File: src/info/worker.ts (hunk 7 above)
Comment (sriGanna) [reply src/info/worker.ts:642]: this should be defined at the top now as a const

File: src/info/info.ts (hunk 9 above)
Comment (sriGanna) [reply src/info/info.ts:399]: yes, we just forgot to use it 😅 I added that to the const hapticPort line below

**Review by jaydeepsingh25 — APPROVED** [review]

**Pull request merged**